Apps/Src/ReadCarCAN.c

```diff
@@ -189,53 +147,75 @@ void Task_ReadCarCAN(void *p_arg)
+        if(status != SUCCESS){
             continue;
         }
 
-        switch(dataBuf.ID){ //we got a message
-            case BPS_TRIP: {
-                // BPS has a fault and we need to enter fault state (probably)
-                if(dataBuf.data[0] == 1){ // If buffer contains 1 for a BPS trip, we should enter a nonrecoverable fault
-
-                    Display_Evac(SOC, SBPV);    // Display evacuation message
 
-                    // Create an exception and assert the error
-                    // kill contactors and enter a nonrecoverable fault
-                    exception_t tripBPSError = {.prio=1, .message="BPS has been tripped", .callback=callback_disableContactors};
-                    assertExceptionError(tripBPSError);
-                }
+        switch(dataBuf.ID){
+            case BPS_TRIP: { // BPS has a fault and we need to enter fault state (probably)
+                
+                // Assert the error to kill contactors, display evacuation message, and enter a nonrecoverable fault
+                assertReadCarCANError(READCARCAN_ERR_BPS_TRIP);
+                
             }
-            case CHARGE_ENABLE: { 
+            case CHARGE_ENABLE:{
 
                 // Restart CAN Watchdog timer
                 OSTmrStart(&canWatchTimer, &err);
                 assertOSError(OS_READ_CAN_LOC, err);
 
+                if(dataBuf.data[0] == 0){
+
+                    // Assert error to disable regen and update saturation in callback function
+                    assertReadCarCANError(READCARCAN_ERR_CHARGE_DISABLE);
 
-                if(dataBuf.data[0] == 0){ // If the buffer doesn't contain 1 for enable, turn off chargeEnable and turn array off
-                    chargingDisable();
-                    updateSaturation(-1); // Update saturation and buffer
+                } else {
 
-                } else { //We got a message of enable with a nonzero value
-                    updateSaturation(1);
+                    // Enable regen and update saturation
+                    updateSaturation(1);                               
+                    
+                    if(arrayIgnitionStatus == true                                                  // Ignition is ON
+                        && chargeMsgSaturation >= SATURATION_THRESHOLD                              // Saturation Threshold has be met
+                        && (Contactors_Get(ARRAY_CONTACTOR)==OFF)                                   // Array Contactor is OFF
+                        && ((OSTmrStateGet(&prechargeDlyTimer, &err) == OS_TMR_STATE_STOPPED)       // Precharge Delay is not running
+                        || (OSTmrStateGet(&prechargeDlyTimer, &err) == OS_TMR_STATE_COMPLETED))){   // or has already completed
+
+                            // NOTE: An assert error for OSTmrStateGet isn't needed here because the 
+                            //       conditional should fail if there is an error. 
```

Comment:
I feel like we should still assert the error since a new error could emerge after running OSTmrStart.